### PR TITLE
Scheduled Fargate Task (to discuss)

### PR DESCRIPTION
**Pull request by mbarton**
Adds a new deployment type wrapping up a container running as a scheduled task in Fargate. It includes some best practice we've discovered working with these in [the Digital Investigations team](https://github.com/guardian/investigations-platform):

- Wrap the invocation in a step function so that we can alarm if the task times out (takes longer than a given amount of time)
- Integrate with Riff-Raff through a Cloudformation deploy, updating the [BuildId parameter](https://github.com/guardian/riff-raff/pull/573)
  - Because we're using Fargate I think we need to pull the container from a registry rather than upload it as part of the artifact
  - We [define the repositories separately](https://github.com/guardian/investigations-platform/blob/b781b8e1ea401243aa26f3cc30f1709ad12aca29/cloudformation/ecr.yaml#L37), granting the Deploy Tools account write access so Teamcity can upload the container

I'd be interested in your thoughts on some things @nicl?

- How could we add additional IAM policy statements to the template?
  - Some of our scrapers access S3, others SSM etc
  - Not language trolling here but if the whole tool were in Typescript it might be easier, as the code could mutate the CDK definition?

It might also be worth scheduling some time to chat through it in person

### CODE REVIEW

**Comment by nicl** [comment]
Great, thansk @mbarton going to take a look at the details and review but very excited by this PR! To quickly answer your two questions:

**On IAM policies:** my original plan was to specify additional policies in the `nest.json` file and attach those. The policies themselves would have to be defined outside of Nest. I still think this approach is worth pursuing and happy to help on this.

**On Go vs Typescript:** I'm hoping that Go support for CDK will come (relatively) soon - https://github.com/orgs/aws/projects/7 and https://github.com/aws/jsii/projects/2. It's being heavily worked on at the moment. At that point, I think the general benefits of Go for a command line tool will dominate. It is slightly awkward for now though for sure.

**Comment by mbarton** [comment]
Thanks Nic.

Makes sense for IAM policies. I've been a bit disappointed with how little CDK helps for those anyway, as far as I can see it doesn't catch any typos or validate that the arns match up to the permissions granted etc.

I didn't realise they were planning proper Go support (beyond shelling out to the Typescript libraries). That's great and worth keeping an eye on.

**Comment by mbarton** [comment]
Something like https://github.com/duo-labs/parliament could be handy for checking the IAM policies, although it's Python so that would be a 3rd language!

File: main.go [hunk 1]
```diff
@@ -176,6 +187,49 @@ func buildArtifact(c config.Config) {
...
+	if err == nil {
+		buildNumber = buildInfo.BuildNumber
+	}
+
+	makeDir(target, c.App)
+	makeDir(target, "cfn")
+
+	containerTag := fmt.Sprintf("%s.dkr.ecr.eu-west-1/%s:%s", c.AccountId, c.App, buildNumber)
+
+	buildOut, err := exec.Command("docker", "build", "-t", containerTag, ".").Output()
```
Comment (nicl) [inline main.go:202]: What happens with the image once created - do we need to do a `docker push` here?

**Review by nicl — COMMENTED** [review]
Happy with the code! Only question was the one about pushing the Docker image to the registry - and how to make this obvious.

In general I was keen for the `build` command to not involve permissions or network calls at all really and just output files. We could look into enabling Riffraff to publish Docker images? It would be more effort but might be a cleaner solution.

**Comment by mbarton** [comment]
Sorry I never had a chance to look at this properly

**Pull request closed without merging**